src/primitives/transaction.h

```diff
@@ -616,21 +655,168 @@ class CTxCertifierLockOut : public CTxCrosschainOut
+    virtual bool CheckOutputsAvailability (CValidationState &state) const = 0;
+    virtual bool CheckSerializedSize      (CValidationState &state) const = 0;
+
+    bool CheckInputsAmount (CValidationState &state) const;
+    bool CheckOutputsAmount(CValidationState &state) const;
+    bool CheckInputsDuplication(CValidationState &state) const;
+    bool CheckInputsInteraction(CValidationState &state) const;
+
+    bool CheckOutputsCheckBlockAtHeightOpCode(CValidationState& state) const;
+
+    bool CheckInputsLimit() const;
+    //END OF CHECK FUNCTIONS
+
+    // Return sum of txouts.
+    virtual CAmount GetValueOut() const;
+
+    int GetComplexity() const { return vin.size()*vin.size(); }
+
+    // Compute modified tx size for priority calculation (optionally given tx size)
+    unsigned int CalculateModifiedSize(unsigned int nTxSize=0) const;
+
+    // Compute priority, given priority of inputs and (optionally) tx size
+    double ComputePriority(double dPriorityInputs, unsigned int nTxSize=0) const;
+
+    // Return sum of JoinSplit vpub_new if supported
+    virtual CAmount GetJoinSplitValueIn() const;
+
+    //-----------------
+    // pure virtual interfaces 
+    virtual bool AcceptTxBaseToMemoryPool(CTxMemPool& pool, CValidationState &state, bool fLimitFree, 
+        bool* pfMissingInputs, bool fRejectAbsurdFee=false) const = 0;
+    virtual void Relay() const = 0;
+    virtual unsigned int GetSerializeSizeBase(int nType, int nVersion) const = 0;
+    virtual std::shared_ptr<const CTransactionBase> MakeShared() const = 0;
+
+    virtual bool CheckFeeAmount(const CAmount& totalVinAmount, CValidationState& state) const = 0;
+
+    virtual bool IsNull() const = 0;
 
+    // return fee amount
+    virtual CAmount GetFeeAmount(const CAmount& valueIn) const = 0;
+
+    // Compute tx size
+    virtual unsigned int CalculateSize() const = 0;
+
+    virtual std::string EncodeHex() const = 0;
+    virtual std::string ToString() const = 0;
+
+    virtual void AddToBlock(CBlock* pblock) const = 0;
+    virtual void AddToBlockTemplate(CBlockTemplate* pblocktemplate, CAmount fee, unsigned int sigops) const = 0;
+
+    virtual bool ContextualCheck(CValidationState& state, int nHeight, int dosLevel) const = 0;
+    virtual bool CheckFinal(int flags = -1) const = 0;
+
+    bool VerifyScript(
+        const CScript& scriptPubKey, unsigned int flags, unsigned int nIn, const CChain* chain,
+        bool cacheStore, ScriptError* serror) const;
+
+    virtual std::shared_ptr<BaseSignatureChecker> MakeSignatureChecker(
+        unsigned int nIn, const CChain* chain, bool cacheStore) const = 0;
+
+    //-----------------
+    // default values for derived classes which do not support specific data structures
+
+    // return false when meaningful only in a block context. As of now only tx coin base returns false
+
+    bool IsCoinBase() const { return GetVin().size() == 1 && GetVin()[0].prevout.IsNull(); }
```

Comment:
make this virtual and implement it differently from CCertificate (always returning false) and CTransaction

Reply:
Done